sdjournal/journal.go

```diff
@@ -84,13 +260,49 @@ func (m *Match) String() string {
 	return m.Field + "=" + m.Value
 }
 
+func (j *Journal) getFunction(name string) (unsafe.Pointer, error) {
+	j.mu.Lock()
+	f, ok := libsystemdFunctions[name]
+	if !ok {
+		var err error
+		f, err = j.lib.GetSymbolPointer(name)
+		if err != nil {
+			return nil, err
+		}
+
+		libsystemdFunctions[name] = f
+	}
+	j.mu.Unlock()
+
+	return f, nil
+}
+
 // NewJournal returns a new Journal instance pointing to the local journal
-func NewJournal() (*Journal, error) {
-	j := &Journal{}
-	r := C.sd_journal_open(&j.cjournal, C.SD_JOURNAL_LOCAL_ONLY)
+func NewJournal() (j *Journal, err error) {
+	h, err := dlopen.GetHandle(libsystemdNames)
+	if err != nil {
+		return nil, err
+	}
+	defer func() {
+		if err != nil {
+			err2 := h.Close()
+			if err2 != nil {
+				err = fmt.Errorf(`%q and "error closing handle: %v"`, err, err2)
```

Comment:
`errwrap.Wrap("error closing handle", errwrap.Wrap(err2, err))`? This means - failed to close the handle, because of some other failure. Maybe that is not perfect (failure in closing handle is not because something else has failed, we are closing the handle because some other error happened and we failed at it).

Until we get some tree-like error handling (to record that something failed because several other things failed) that's the best we can get probably.